core/src/main/java/org/apache/iceberg/BaseFile.java

```diff
@@ -504,6 +508,27 @@ private static Map<Integer, ByteBuffer> toReadableByteBufferMap(Map<Integer, Byt
     }
   }
 
+  private static <TypeT> Map<Integer, TypeT> filterColumnsStats(
+      Map<Integer, TypeT> map, Set<Integer> columnIds) {
+    if (columnIds == null || columnIds.isEmpty()) {
+      return SerializableMap.copyOf(map);
+    }
+
+    if (map == null) {
+      return null;
+    }
+
+    Map<Integer, TypeT> filtered = Maps.newHashMapWithExpectedSize(columnIds.size());
+    for (Integer columnId : columnIds) {
+      TypeT value = map.get(columnId);
+      if (value != null) {
```

Comment:
Yes, I found this in some tests, that the map could contain null values

Reply:
If there are null values in lower_bound or upper_bound, why do we want to filter them out? shouldn't we keep the same behavior when copying stats for selected columns.

Reply:
sure. but could lower_bound and upper_bound have null value? how do we distinguish btw these two scenario?

e.g. here is the description for lower_bound
```
Lower bound for the non-null, non-NaN values in the partition field, or null if all values are null or NaN [2]
```

Reply:
I did some testing with Parquet files:
- Used `TestParquet` to check what `Metrics` do we generate when we have only `null` values:
```

  @Test
  public void testNullMetricsForEmptyColumns() throws IOException {
    Schema schema = new Schema(optional(1, "longCol", Types.LongType.get()),
        optional(2, "nullCol", Types.LongType.get()));

    File file = createTempFile(temp);

    List<GenericData.Record> records = Lists.newArrayListWithCapacity(1);
    org.apache.avro.Schema avroSchema = AvroSchemaUtil.convert(schema.asStruct());

    GenericData.Record smallRecord = new GenericData.Record(avroSchema);
    smallRecord.put("longCol", 1L);
    records.add(smallRecord);


    write(
        file,
        schema,
        Collections.emptyMap(),
        ParquetAvroWriter::buildWriter,
        records.toArray(new GenericData.Record[] {}));

    InputFile inputFile = Files.localInput(file);
    Metrics metrics = ParquetUtil.fileMetrics(inputFile, MetricsConfig.getDefault());
    assertThat(metrics.nullValueCounts()).hasSize(2);
    assertThat(metrics.lowerBounds()).hasSize(1).containsKey(1);
    assertThat(metrics.upperBounds()).hasSize(1).containsKey(1);
  }
```
Notice, that the size of the `lowerBounds` and the `upperBounds` map is 1. There is no `null` value for column '2'

- Just to double check, I have tested how we read files with missing column metrics (like the ones created above). I have used `TestManifestReaderStats` for this:
```
  @Test
  public void testReadFileWithMissingStatsIncludesFullStats() throws IOException {
    Metrics missingStats =
        new Metrics(
            3L, null,
            ImmutableMap.of(3, 3L, 4, 4L),
            ImmutableMap.of(3, 3L, 4, 0L),
            ImmutableMap.of(3, 0L, 4, 1L),
            ImmutableMap.of(4, Conversions.toByteBuffer(Types.StringType.get(), "a")),
            ImmutableMap.of(4, Conversions.toByteBuffer(Types.StringType.get(), "a")));
    DataFile fileWithMissingStats =
        DataFiles.builder(SPEC)
            .withPath(FILE_PATH)
            .withFileSizeInBytes(10)
            .withPartitionPath("data_bucket=0") // easy way to set partition data for now
            .withRecordCount(3)
            .withMetrics(missingStats)
            .build();
    ManifestFile manifest = writeManifest(1000L, fileWithMissingStats);
    try (ManifestReader<DataFile> reader = ManifestFiles.read(manifest, FILE_IO)) {
      CloseableIterable<ManifestEntry<DataFile>> entries = reader.entries();
      ManifestEntry<DataFile> entry = entries.iterator().next();
      DataFile dataFile = entry.file();
      Assert.assertEquals(3, dataFile.recordCount());
      assertThat(dataFile.lowerBounds()).hasSize(1).containsKey(4);
      assertThat(dataFile.upperBounds()).hasSize(1).containsKey(4);
    }
  }
```
Notice, that the size of the `lowerBounds` and the `upperBounds` map is 1. There is no `null` value for column '3'

So I think the actual phrasing of the documentation is just not clean enough, as the map does not contain `null` values, it just does not contain the keys. While putting `null` value to a map is certainly possible, it should be used only in very specific cases, as it is very error-prone (see our discussion above 😄). If our goal would be to identify the case where all of the values in the given column are `null`, we can use the other stats to archive the result.

Reply:
thanks for the unit test code. I am still not comfortable of making the assumption as the Iceberg spec clearly said null stats value is possible.
```
Lower bound for the non-null, non-NaN values in the partition field, or null if all values are null or NaN [2]
```

Unless we are going to clarify/update the spec, I feel it is safer to do `map.containsKey(key)`.

Maybe @RussellSpitzer and @rdblue can chime in here.

Reply:
Changed to ensure that even `null` values are copied if we accept the cost of possible double fetching the entries from the map.

Still not sure that this is necessary, but I will be on PTO next week, so wanted to have it here as a possibility

Reply:
thanks.  double fetching is probably ok since the columnsToKeepStats is likely not big and Map get is a O(1) anyway.

The alternative is to start a discussion and revisit the spec or Parquet implementation.